lychee-lib/src/checker.rs

```diff
@@ -0,0 +1,67 @@
+use crate::{
+    chain::{ChainResult, Chainable},
+    retry::RetryExt,
+    Status,
+};
+use async_trait::async_trait;
+use http::StatusCode;
+use reqwest::Request;
+use std::{collections::HashSet, time::Duration};
+
+#[derive(Debug, Clone)]
+pub(crate) struct Checker {
+    retry_wait_time: Duration,
+    max_retries: u64,
+    reqwest_client: reqwest::Client,
+    accepted: Option<HashSet<StatusCode>>,
+}
+
+impl Checker {
+    pub(crate) const fn new(
+        retry_wait_time: Duration,
+        max_retries: u64,
+        reqwest_client: reqwest::Client,
+        accepted: Option<HashSet<StatusCode>>,
+    ) -> Self {
+        Self {
+            retry_wait_time,
+            max_retries,
+            reqwest_client,
+            accepted,
+        }
+    }
+
+    /// Retry requests up to `max_retries` times
+    /// with an exponential backoff.
+    pub(crate) async fn retry_request(&self, request: Request) -> Status {
+        let mut retries: u64 = 0;
+        let mut wait_time = self.retry_wait_time;
+
+        let mut status = self.check_default(request.try_clone().unwrap()).await; // TODO: try_clone
+        while retries < self.max_retries {
+            if status.is_success() || !status.should_retry() {
+                return status;
+            }
+            retries += 1;
+            tokio::time::sleep(wait_time).await;
+            wait_time = wait_time.saturating_mul(2);
+            status = self.check_default(request.try_clone().unwrap()).await; // TODO: try_clone
```

Comment:
@mre There is no simple alternative to `try_clone().unwrap()` I could come up with. Or do you have another idea. I guess the only alternative would be to change `RequestChain` to not use `reqwest::Request` but our own wrapped type as we briefly touched on in a previous discussion.

I think the current approach is fine since we don't make use of the `stream` feature of `reqwest`. Do you agree?

```
reqwest = { version = "0.11.23", default-features = false, features = [
    "gzip",
    "trust-dns",
    "cookies",
] }
```

Reply:
Yeah, it's fine. As discussed, let's just add a `SAFETY` statement here, so we mention why it will always work as long as we don't add `stream`. 😃